feat(zcash): initial commit

MrCyjaneK opened the pull request:
# Description

Zcash c:

# Pull Request - Checklist  

- [ ] Initial Manual Tests Passed
- [ ] Double check modified code and verify it with the feature/task requirements
- [ ] Format code
- [ ] Look for code duplication
- [ ] Clear naming for variables and methods
- [ ] Manual tests in accessibility mode (TalkBack on Android) passed

OmarHatem28 commented on cw_core/lib/crypto_currency.dart:
```diff
@@ -186,7 +186,7 @@ class CryptoCurrency extends EnumerableItem<int> with Serializable<int> implemen
   static const usdcsol = CryptoCurrency(title: 'USDC', tag: 'SOL', fullName: 'USDC Coin', raw: 41, name: 'usdcsol', iconPath: 'assets/images/crypto/usdc.webp', decimals: 6);
   static const zaddr = CryptoCurrency(title: 'ZZEC', tag: 'ZEC', fullName: 'Shielded Zcash', raw: 42, name: 'zaddr', iconPath: 'assets/images/zec_icon.png', decimals: 8);
-  static const zec = CryptoCurrency(title: 'TZEC', tag: 'ZEC', fullName: 'Transparent Zcash', raw: 43, name: 'zec', iconPath: 'assets/images/zec_icon.png', decimals: 8);
+  static const tzec = CryptoCurrency(title: 'tZEC', tag: 'tZEC', fullName: 'Transparent Zcash', raw: 43, name: 'zec', iconPath: 'assets/images/zec_icon.png', decimals: 8);
```
should be added to the `all` list

MrCyjaneK replied:
https://github.com/cake-tech/cake_wallet/blob/942feb662709dffe69a6d695760d26cfb076bb54/cw_core/lib/crypto_currency.dart#L66 It's there

OmarHatem28 replied:
was talking about tzec, but I added it anyway

OmarHatem28 commented on cw_core/lib/node.dart:
```diff
@@ -181,6 +196,7 @@ class Node extends HiveObject with Keyable {
           return requestZanoNode();
         case WalletType.decred:
           return requestDecredNode();
+        case WalletType.zcash:
```
why no pinging?

MrCyjaneK replied:
Fixed now, was wip

MrCyjaneK committed:
feat(zcash): initial commit
- initial work on cw_zcash
  - use single .db for everything zcash to speed up sync and scanning
  - implement silent autoshielding of non-orchard funds
  - use FFI for everything to fetch the data
  - add support for passphrase by adding extra word to seed phrase
  - enforce lints on CI for cw_zcash directory
- use correct zcash name, and call transparent zcash tZEC
- minor cleanups outsite of cw_zcash
  - gitmodules, gitignore
  - db debug
  - fix broken addressPageType in WalletInfo
  - enable debug options in kDebugMode and kProfileMode

MrCyjaneK committed:
feat: T address rotation (and cache) wip: minor fixes

MrCyjaneK committed:
wip: shield tx list

MrCyjaneK committed:
fix fiat amounts fix derrivation of rotation account addresses screen
for zcash block height fix T address shields improvements Amount in
disposable T shields UI improvements
show pending outgoing tx in tx history

OmarHatem28 commented on pubspec_base.yaml:
```diff
@@ -240,9 +240,12 @@ flutter:
     - assets/faq/
     - assets/animation/
+    - scripts/zcash_lib/assets/sapling-spend.params
+    - scripts/zcash_lib/assets/sapling-output.params
```
what's that?

MrCyjaneK replied:
Requirement for zcash sapling to work

OmarHatem28 commented on lib/view_model/wallet_keys_view_model.dart:
```diff
@@ -270,6 +271,8 @@ abstract class WalletKeysViewModelBase with Store {
       case WalletType.dogecoin:
         return 'dogecoin-wallet';
+      case WalletType.zcash:
+        return 'zcash-wallet';
```
you missed adding this scheme to android manifest base and info.plist

MrCyjaneK committed:
fix null check on passphrase page fix tx showing as pending rename auto
shield txs fixed send all to include fee use rotating address in
addressForExchange wrap all calls of WarpApi.getLatestHeight in try
catch to prevent leaking errors

other minor fixes from notion

MrCyjaneK committed:
wallet group debug cache sent tx address wallet T address rotation fixes
automatically pick the fee for auto-shield

OmarHatem28 commented on lib/reactions/on_wallet_sync_status_change.dart:
```diff
@@ -31,6 +33,12 @@ void startWalletSyncStatusChangeReaction(
+          await walletListViewModel.updateList();
+        } catch (e) {
+          printV("Error refreshing wallet list after sync: $e");
+        }
```
why is this?

MrCyjaneK replied:
It wasn't properly refreshing the list, **probably** can be removed, will check that later

OmarHatem28 commented on lib/view_model/settings/other_settings_view_model.dart:
is this not used or something is missing here?

MrCyjaneK replied:
It used to have zec db debug option, but I've removed it because it was locking the db and causing a little mess ther

OmarHatem28 commented on scripts/lint.sh:
```diff
@@ -0,0 +1,18 @@
+for i in cw_zcash;
+do
+    dart fix --apply $i/
+    dart format --line-length=120 $i/
```
to match the rest of the project and the team configuration
```suggestion
    dart format --line-length=100 $i/
```

OmarHatem28 commented on cw_zcash/lib/src/zcash_transaction_info.dart:
```diff
@@ -0,0 +1,100 @@
+    _destinationAddressMap[txId] = address;
+    final pfwt = await pathForWalletTypeDir(type: WalletType.zcash);
+    final f = File(p.join(pfwt, "sent-tx-map.json"));
+    f.writeAsStringSync(json.encode(f));
```
shouldn't this be like this?
```suggestion
    f.writeAsStringSync(json.encode(_destinationAddressMap));
```

MrCyjaneK replied:
yes

OmarHatem28 reviewed (commented):
7 files left to review, but here is my comments so far

MrCyjaneK committed:
fix T address cache deserialization fix autoshield icon fix autoshield
text getCachedDestinationAddress fix send address cache

MrCyjaneK committed:
address comments from review

MrCyjaneK committed:
wip: iOS fix

OmarHatem28 committed:
fix swap

OmarHatem28 committed:
Allow cakepay payment from zcash wallet

MrCyjaneK committed:
fix: ios (I'm stupid) fix: reorder addresses to make orchard default
fix: send all balance

MrCyjaneK committed:
5.6.6 bump fix: ios apple review add transport error to ignored
exceptions

MrCyjaneK committed:
fix: leaving dust behind when sweeping all

OmarHatem28 commented on android/app/src/main/AndroidManifestBase.xml:
```diff
@@ -112,6 +112,7 @@
                 <data android:scheme="arbitrum" />
                 <data android:scheme="arbitrum-wallet" />
                 <data android:scheme="arbitrum_wallet" />
+                <data android:scheme="zcash-wallet" />
```
```suggestion
                <data android:scheme="zcash" />
                <data android:scheme="zcash-wallet" />
```

OmarHatem28 committed:
Change wallets order

MrCyjaneK committed:
fix: reflect balance of pending out txs fix: restore height support

OmarHatem28 commented on ios/Runner/InfoBase.plist:
```diff
@@ -342,6 +342,16 @@
                 <string>arbitrum-wallet</string>
             </array>
         </dict>
+        <dict>
+            <key>CFBundleTypeRole</key>
+            <string>Viewer</string>
+            <key>CFBundleURLName</key>
+            <string>zcash-wallet</string>
+            <key>CFBundleURLSchemes</key>
+            <array>
+                <string>zcash-wallet</string>
+            </array>
+        </dict>
```
```suggestion
        <dict>
            <key>CFBundleTypeRole</key>
            <string>Viewer</string>
            <key>CFBundleURLName</key>
            <string>zcash</string>
            <key>CFBundleURLSchemes</key>
            <array>
                <string>zcash</string>
            </array>
        </dict>
        <dict>
            <key>CFBundleTypeRole</key>
            <string>Viewer</string>
            <key>CFBundleURLName</key>
            <string>zcash-wallet</string>
            <key>CFBundleURLSchemes</key>
            <array>
                <string>zcash-wallet</string>
            </array>
        </dict>
```

OmarHatem28 commented on lib/zcash/cw_zcash.dart:
```diff
@@ -0,0 +1,2263 @@
+  }
+}
+
+const wordList = [
```
let's remember to just have a single file with bip39 words and use it everywhere instead

MrCyjaneK committed:
linux support (disabled)

MrCyjaneK committed:
fix: date picker for zcash on restore screen fix: build script for linux
match WarpApi.warpSync args with upstream

MrCyjaneK committed:
fix: linter script

serhii-bor committed:
ensure all exchange providers support ZEC (#2802)

* ensure all exchange providers support ZEC

* remove zaddr and tzec currencies and refactor providers

* Update lib/exchange/provider/near_Intents_exchange_provider.dart

* add back old zcash currencies for backward compatibility, thanks Czarek

* minor fix [skip ci]

---------

Co-authored-by: Omar Hatem <omarh.ismail1@gmail.com>

OmarHatem28 committed:
Update android/app/src/main/AndroidManifestBase.xml [skip ci]

OmarHatem28 committed:
Update ios/Runner/InfoBase.plist [skip ci]

MrCyjaneK committed:
fix: remove unused code fix: macos fix: disable on linux and macos

OmarHatem28 committed:
update build numbers [skip ci]

OmarHatem28 commented on cw_zcash/lib/src/zcash_taddress_rotation.dart:
```diff
@@ -0,0 +1,421 @@
+    if (acc == null) {
+      printV("Nothing found");
+      return null;
+    }
```
isn't `seed` a string, and the map is of int key, wouldn't this always return null

MrCyjaneK committed:
feat: use zkool
fix: hack zkool to use 28.2NDK
drop T addr rotation for now
essentially rewrite everything.

testers: do not update, create backup first on previous version
there's no migration step here.

DO NOT TEST

OmarHatem28 commented:
@MrCyjaneK I already took a branch from the last commit before you added zkool and resolved the conflicts with dev and merged it into dev, so just base your latest commit on latest dev or cherry pick it on a new branch
